pkg/kubelet/cm/dra/manager.go

```diff
@@ -302,34 +335,58 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
-				return fmt.Errorf("failed to get DRA Plugin client for plugin name %s, err=%+v", pluginName, err)
+			batches[pluginName] = append(batches[pluginName],
+				&drapb.Claim{
+					Namespace:      resourceClaim.Namespace,
+					Uid:            string(resourceClaim.UID),
+					Name:           resourceClaim.Name,
+					ResourceHandle: resourceHandle.Data,
+				})
+		}
+		claimInfos[resourceClaim.UID] = claimInfo
+	}
+
+	// Check for errors. If there is any error, processing gets aborted.
+	// We could try to continue, but that would make the code more complex.
+	for pluginName, claims := range batches {
+		// Call NodeUnprepareResources RPC for all resource handles.
+		client, err := dra.NewDRAPluginClient(pluginName)
+		if err != nil {
+			return fmt.Errorf("failed to get DRA Plugin client for plugin name %s: %v", pluginName, err)
+		}
+		response, err := client.NodeUnprepareResources(context.Background(), &drapb.NodeUnprepareResourcesRequest{Claims: claims})
+		if err != nil {
+			// General error unrelated to any particular claim.
+			return fmt.Errorf("NodeUnprepareResources failed: %v", err)
+		}
+
+		for claimUID, result := range response.Claims {
+			reqClaim := lookupClaimRequest(claims, claimUID)
+			if reqClaim == nil {
+				return fmt.Errorf("NodeUnprepareResources returned result for unknown claim UID %s", claimUID)
 			}
-			response, err := client.NodeUnprepareResource(
-				context.Background(),
-				claimInfo.Namespace,
-				claimInfo.ClaimUID,
-				claimInfo.ClaimName,
-				resourceHandle.Data)
-			if err != nil {
-				return fmt.Errorf(
-					"NodeUnprepareResource failed, pod: %s, claim UID: %s, claim name: %s, resource handle: %s, err: %+v",
-					pod.Name, claimInfo.ClaimUID, claimInfo.ClaimName, resourceHandle.Data, err)
+			if result.Error != "" {
+				return fmt.Errorf("NodeUnprepareResources failed for claim %s/%s: %s", reqClaim.Namespace, reqClaim.Name, err)
 			}
-			klog.V(3).InfoS("NodeUnprepareResource succeeded", "response", response)
 		}
 
-		// Delete last pod UID only if all NodeUnprepareResource calls succeed.
+		unfinished := len(claims) - len(response.Claims)
+		if unfinished != 0 {
+			return fmt.Errorf("NodeUnprepareResources left out %d claims", unfinished)
+		}
+	}
+
+	// No errors, time to remove the pod reference from all claims that were involved.
+	for _, claimInfo := range claimInfos {
+		// Delete last pod UID only if unprepare succeed.
 		// This ensures that the status manager doesn't enter termination status
 		// for the pod. This logic is implemented in
 		// m.PodMightNeedToUnprepareResources and claimInfo.hasPodReference.
 		claimInfo.deletePodReference(pod.UID)
 		m.cache.delete(claimInfo.ClaimName, pod.Namespace)
 
-		// Checkpoint to reduce redundant calls to NodeUnPrepareResource() after a kubelet restart.
-		err = m.cache.syncToCheckpoint()
+		// Checkpoint to reduce redundant calls to NodeUnPrepareResources() after a kubelet restart.
+		err := m.cache.syncToCheckpoint()
```

Comment:
If we batch the pod reference deletions at the end like this, then I think the call to `syncToCheckpoint ()` within this loop can simply be removed. We call it again just after the loop and without a bunch of gRPC calls between the checkpoints as before (which is why we wanted to checkpoint early in case one of the `NodeUnprepareResource()` calls failed).

Reply:
I think the `syncToCheckpoint` at the end of each function also was redundant. Doing it once after each gRPC call should be enough. Changed.

Reply:
... except when the entire call was skipped. Then the final sync is needed.

Reply:
I actually meant the other way around -- the final one is necessary -- the intermediate ones are not.

Reply:
So prune them even further, i.e. do it potentially after multiple `NodePrepareResources` calls? That can happen when there are multiple plugins involved.